### ISSUE TITLE
Banner-based issue

### ISSUE THREAD

**Issue by Sunconure11** [Paperdomo101/lightstones#10]
https://github.com/Darkhax-Minecraft/AdditionalBanners/issues/27

@Paperdomo101 We were discussing a Rahkshi banner on that one forum, but we may want to deal with this first.

**Comment by Sunconure11** [comment]
https://pastebin.com/Lw3kuXsq

**Comment by Sunconure11** [comment]
IIRC, @darkhax did outline to me a safer method. Maybe he can post it here?

**Comment by Paperdomo101** [comment]
Alright, let me see if I've got this right, and please correct me if I'm wrong.

- There's a mixin collision with the mod Additional Banners, and it's because of the implementation of custom banners in Lightstones.

- BannerPlusPlus is currently outdated, and unless there is another mod that enables custom banner registry, there's no library mod to solve this problem.

- I assume the problem persists in the 1.6.1 version of Lightstones? I realized the banner client registry function wasn't being called under the 'client' entrypoint in the fabric.mod.json, so I moved its initialization call from the main Lightstones class to LightstonesClient. If it's a mixin problem though, I doubt this would make any difference.

If @Darkhax has a method that will avoid the collision we should definitely implement it.

**Comment by Sunconure11** [comment]
The banner code inside the mod is an implementation of bannerplusplus.

Darkhax’s additional banners has a safer implementation of banner code, as opposed to my port of bannerplusplus

**Comment by Sunconure11** [comment]
I'd talk to DarkHax to guide you through this.

Most MC modding stuff is discussed on Discord, so add him if you have one. I tried to see if you had one but I couldn't find it.

Here is Dark's:

Darkhax#6676

Due to some mental health issues, I will let him guide you through fixing my BS. I am going to absent from most coding until sometime after the 8th.

**Comment by Darkhax** [comment]
> There's a mixin collision with the mod Additional Banners, and it's because of the implementation of custom banners in Lightstones.

I am not sure if this is actually the case. I still haven't seen a crash log that shows the mod being incompatible with mine. All of the crash reports are stuff like client code being loaded on the server or mixin errors from environments where my mod is not even installed.

> If @Darkhax has a method that will avoid the collision we should definitely implement it.

For 1.18.2 people have just been using the enum extender from [FabricASM](https://github.com/PlanetTeamSpeakk/Fabric-ASM). You can see my code for this [here](https://github.com/Darkhax-Minecraft/AdditionalBanners/blob/1.18.2/Fabric/src/main/java/net/darkhax/additionalbanners/AdditionalBannersEarlyRiser.java). This is the same approach used by Twilight Forest and can be found [here](https://github.com/TeamTwilight/twilightforest-fabric/blob/b282d180f6a84465a5596823c34e4035f66aaba2/src/main/java/twilightforest/asm/TFASM.java#L30-L40). This approach is roughly analogous to how banner patterns are added on Forge.

The main disadvantage of this approach is that you can only add stencil-bound patterns and not the generally available patterns. This isn't an issue for most mods. In 1.19 Mojang has moved this to a proper registry rather than an enum and use tags to determine how the pattern is available in the loom.

**Comment by Sunconure11** [comment]
Doing some looking at Twilight Forest now, it is definitely not the cause, like that person thought.

**Comment by Sunconure11** [comment]
Okay, so Twilight Forest is actually doing something unsafe, after all.

**Comment by Sunconure11** [comment]
Alright, so Twilight Forest is handling things safely. I was wrong.

Edit: I think it may need to be investigated in further detail. It ran fine with Lightstones in an isolated environment, but shat itself in a more modded environment.

Edit 2: Error was unrelated. Twilight Forest is doing things right.

**Comment by Paperdomo101** [comment]
@Darkhax Thank you for your time and sharing your implementation, I appreciate it.
As of the currently unpublished version of Lightstones (1.6.1), our mods do not conflict, as you can see here:
https://i.imgur.com/vBdiKTl.png
https://i.imgur.com/BlSJBCc.png

@Sunconure11 I believe the cause was (as I have mentioned to you before) initializing client banner registry code under the server init function, which I fixed by moving it to the client init function. Have all these tests been done with Lightstones 1.6.0? Regardless, I will publish v1.6.1 on Curseforge shortly, which will (hopefully) have this issue fixed.

Edit: Just published Lightstones 1.6.1

**Comment by Sunconure11** [comment]
The main issue is that we are using mixins to add the banners.

I'm still getting the issue, so I'll see if I can redirect you to the folders that contain all of that stuff

**Comment by Sunconure11** [comment]
https://github.com/Paperdomo101/lightstones/tree/master/src/main/java/paperdomo101/lightstones/api/banner

https://github.com/Paperdomo101/lightstones/tree/master/src/main/java/paperdomo101/lightstones/mixin/block

https://github.com/Paperdomo101/lightstones/tree/master/src/main/java/paperdomo101/lightstones/mixin/client

https://github.com/Paperdomo101/lightstones/blob/master/src/main/java/paperdomo101/lightstones/registry/LightstonesBannerPatterns.java

https://github.com/Paperdomo101/lightstones/search?q=banner

The method Dark is using does not call upon mixins, and as such, has less compat issues. Furthermore, the license on his mod lets you take code from it, if you need to use it. I would strongly advise switching over from all of this to that method.

**Comment by Sunconure11** [comment]
@Paperdomo101

**Comment by Sunconure11** [comment]
Tl;dr we need a rewrite of banners, but I'm waiting on an appointment with my psychiatrist before I do anything here, so it'd be up to you.

**Comment by Paperdomo101** [comment]
Okay, I will get to work on this as soon as I have some time. I look forward to 1.19 when this problem will be magicked away. Please take it easy @Sunconure11, hope your appointment goes well.

**Comment by Sunconure11** [comment]
https://github.com/Darkhax-Minecraft/AdditionalBanners/blob/1.18.2/Fabric/src/main/java/net/darkhax/additionalbanners/AdditionalBannersEarlyRiser.java

https://github.com/TeamTwilight/twilightforest-fabric/blob/b282d180f6a84465a5596823c34e4035f66aaba2/src/main/java/twilightforest/asm/TFASM.java#L30-L40

https://github.com/PlanetTeamSpeakk/Fabric-ASM

**Comment by itsdinkd** [comment]
1.6.1 did not fix the issue. like mentioned in my past issue, this problem never occurred until I added twilight forest. The server doesn't crash ( at least what I wrote the issue on), but client does. you cannot launch the game.


```java
[15:32:21] [main/ERROR]: Minecraft has crashed!
net.fabricmc.loader.impl.FormattedException: java.lang.ExceptionInInitializerError
	at Not Enough Crashes deobfuscated stack trace.(1.18.2+build.3) ~[?:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:462) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) [fabric-loader-0.14.6.jar:?]
Caused by: java.lang.ExceptionInInitializerError
	at net.minecraft.world.gen.carver.CarverDebugConfig.<clinit>(CarverDebugConfig:9) ~[?:?]
	at net.minecraft.world.gen.carver.CarverConfig.method_33968(CarverConfig:16) ~[?:?]
	at com.mojang.serialization.codecs.RecordCodecBuilder.mapCodec(RecordCodecBuilder.java:76) ~[datafixerupper-4.1.27.jar:?]
	at net.minecraft.world.gen.carver.CarverConfig.<clinit>(CarverConfig:12) ~[?:?]
	at net.minecraft.world.gen.carver.Carver.<clinit>(Carver:33) ~[?:?]
	at net.minecraft.world.gen.carver.ConfiguredCarvers.<clinit>(ConfiguredCarvers:19) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30572(BuiltinRegistries:52) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30566(BuiltinRegistries:86) ~[?:?]
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:721) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.<clinit>(BuiltinRegistries:85) ~[?:?]
	at net.minecraft.util.registry.Registry.<clinit>(Registry:291) ~[?:?]
	at net.minecraft.Bootstrap.initialize(Bootstrap:49) ~[?:?]
	at net.minecraft.client.main.Main.main(Main:153) ~[fabric-loader-0.14.6-1.18.2.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:460) ~[fabric-loader-0.14.6.jar:?]
	... 2 more
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1726 failed
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:418) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:322) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:145) ~[fabric-loader-0.14.6.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at net.minecraft.block.Blocks.<clinit>(Blocks:791) ~[?:?]
	at net.minecraft.world.gen.carver.CarverDebugConfig.<clinit>(CarverDebugConfig:9) ~[?:?]
	at net.minecraft.world.gen.carver.CarverConfig.method_33968(CarverConfig:16) ~[?:?]
	at com.mojang.serialization.codecs.RecordCodecBuilder.mapCodec(RecordCodecBuilder.java:76) ~[datafixerupper-4.1.27.jar:?]
	at net.minecraft.world.gen.carver.CarverConfig.<clinit>(CarverConfig:12) ~[?:?]
	at net.minecraft.world.gen.carver.Carver.<clinit>(Carver:33) ~[?:?]
	at net.minecraft.world.gen.carver.ConfiguredCarvers.<clinit>(ConfiguredCarvers:19) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30572(BuiltinRegistries:52) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30566(BuiltinRegistries:86) ~[?:?]
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:721) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.<clinit>(BuiltinRegistries:85) ~[?:?]
	at net.minecraft.util.registry.Registry.<clinit>(Registry:291) ~[?:?]
	at net.minecraft.Bootstrap.initialize(Bootstrap:49) ~[?:?]
	at net.minecraft.client.main.Main.main(Main:153) ~[fabric-loader-0.14.6-1.18.2.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:460) ~[fabric-loader-0.14.6.jar:?]
	... 2 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:413) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:322) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:145) ~[fabric-loader-0.14.6.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at net.minecraft.block.Blocks.<clinit>(Blocks:791) ~[?:?]
	at net.minecraft.world.gen.carver.CarverDebugConfig.<clinit>(CarverDebugConfig:9) ~[?:?]
	at net.minecraft.world.gen.carver.CarverConfig.method_33968(CarverConfig:16) ~[?:?]
	at com.mojang.serialization.codecs.RecordCodecBuilder.mapCodec(RecordCodecBuilder.java:76) ~[datafixerupper-4.1.27.jar:?]
	at net.minecraft.world.gen.carver.CarverConfig.<clinit>(CarverConfig:12) ~[?:?]
	at net.minecraft.world.gen.carver.Carver.<clinit>(Carver:33) ~[?:?]
	at net.minecraft.world.gen.carver.ConfiguredCarvers.<clinit>(ConfiguredCarvers:19) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30572(BuiltinRegistries:52) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30566(BuiltinRegistries:86) ~[?:?]
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:721) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.<clinit>(BuiltinRegistries:85) ~[?:?]
	at net.minecraft.util.registry.Registry.<clinit>(Registry:291) ~[?:?]
	at net.minecraft.Bootstrap.initialize(Bootstrap:49) ~[?:?]
	at net.minecraft.client.main.Main.main(Main:153) ~[fabric-loader-0.14.6-1.18.2.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:460) ~[fabric-loader-0.14.6.jar:?]
	... 2 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector addBppLoomPatternCondition(Lnet/minecraft/class_3915;)I in lightstones.mixins.json:block.LoomContainerMixin from mod lightstones failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap lightstones-refmap.json
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:468) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1385) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:1052) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:400) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:325) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.11.4+mixin.0.8.5.jar:0.11.4+mixin.0.8.5]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:413) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:322) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.14.6.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:145) ~[fabric-loader-0.14.6.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at net.minecraft.block.Blocks.<clinit>(Blocks:791) ~[?:?]
	at net.minecraft.world.gen.carver.CarverDebugConfig.<clinit>(CarverDebugConfig:9) ~[?:?]
	at net.minecraft.world.gen.carver.CarverConfig.method_33968(CarverConfig:16) ~[?:?]
	at com.mojang.serialization.codecs.RecordCodecBuilder.mapCodec(RecordCodecBuilder.java:76) ~[datafixerupper-4.1.27.jar:?]
	at net.minecraft.world.gen.carver.CarverConfig.<clinit>(CarverConfig:12) ~[?:?]
	at net.minecraft.world.gen.carver.Carver.<clinit>(Carver:33) ~[?:?]
	at net.minecraft.world.gen.carver.ConfiguredCarvers.<clinit>(ConfiguredCarvers:19) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30572(BuiltinRegistries:52) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.method_30566(BuiltinRegistries:86) ~[?:?]
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:721) ~[?:?]
	at net.minecraft.util.registry.BuiltinRegistries.<clinit>(BuiltinRegistries:85) ~[?:?]
	at net.minecraft.util.registry.Registry.<clinit>(Registry:291) ~[?:?]
	at net.minecraft.Bootstrap.initialize(Bootstrap:49) ~[?:?]
	at net.minecraft.client.main.Main.main(Main:153) ~[fabric-loader-0.14.6-1.18.2.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:460) ~[fabric-loader-0.14.6.jar:?]
	... 2 more

```

**Comment by Sunconure11** [comment]
@Paperdomo101 Requesting a PR to my fork so I can implement the banner fix, due to conflicting files

**Comment by Sunconure11** [comment]
On second thought, I seem to be getting at it, but the dependency for Chocohead's ASM does not want to load properly in the workspace. Like, it won't download.

**Comment by Sunconure11** [comment]
@itsdinkd I've got something coming up, mind testing when I do post it?

**Comment by Sunconure11** [comment]
[lightstones.zip](https://github.com/Paperdomo101/lightstones/files/8870337/lightstones.zip)

Try this build of Lightstones.

**Comment by Sunconure11** [comment]
Did some testing, looks like it does not work as intended. I am going to talk to who is behind the port for Banners Plus Plus for a solution, since there is a 1.18 build.

**Comment by TheEpicBlock** [comment]
Lightstones reimplements the mixins of bannerpp, which causes a conflict with the actual bannerpp library. A solution would be to jij bannerpp instead of copying the mixins

**Comment by Sunconure11** [comment]
> Lightstones reimplements the mixins of bannerpp, which causes a conflict with the actual bannerpp library. A solution would be to jij bannerpp instead of copying the mixins

As we had discussed on Discord, the banner issue is related to how other mods are doing banners. Let me know when you’re able to get to it so I can get this solved, once and for all.

**Comment by TheEpicBlock** [comment]
The issue Sunconure is talking about is unrelated to this one. This one is purely because mixin's are copied verbatim, which causes issues with the redirect. Bannerspp' implementation shouldn't conflict with the enum extension that other mods are doing.

**Comment by Sunconure11** [comment]
@Paperdomo101 Requesting a build for 1.6.2 be published in the mean time, since it switches how Banners++ is called.

**Comment by Sunconure11** [comment]
I think we can call this issue fixed. However, if need-be, come back to this, for reference. I'm gonna try and port Ars Theurgia sometime after the 20th.

**Comment by itsdinkd** [comment]
Was it tested with Twilight Forest? the main issue was that it never worked with TF (Additional Banners was not in my pack, ever.)

**Comment by Sunconure11** [comment]
Yes, but if you want, you can test yourself just to verify our results
